src/commonMain/kotlin/fr/acinq/lightning/payment/OfferManager.kt

```diff
@@ -0,0 +1,142 @@
+import fr.acinq.lightning.channel.states.Offline
+import fr.acinq.lightning.channel.states.Syncing
+import fr.acinq.lightning.crypto.RouteBlinding
+import fr.acinq.lightning.io.PayOffer
+import fr.acinq.lightning.logging.MDCLogger
+import fr.acinq.lightning.message.OnionMessages
+import fr.acinq.lightning.message.OnionMessages.Destination
+import fr.acinq.lightning.message.OnionMessages.IntermediateNode
+import fr.acinq.lightning.message.OnionMessages.buildMessage
+import fr.acinq.lightning.utils.currentTimestampMillis
+import fr.acinq.lightning.utils.currentTimestampSeconds
+import fr.acinq.lightning.utils.msat
+import fr.acinq.lightning.utils.toByteVector
+import fr.acinq.lightning.wire.*
+import kotlin.math.max
+
+sealed class OnionMessageAction {
+    data class SendMessage(val message: OnionMessage): OnionMessageAction()
+    data class PayInvoice(val payOffer: PayOffer, val invoice: Bolt12Invoice, val payerKey: PrivateKey): OnionMessageAction()
+    data class ReportPaymentFailure(val payOffer: PayOffer, val failure: OfferPaymentFailure): OnionMessageAction()
+}
+
+private data class PendingInvoiceRequest(val payOffer: PayOffer, val payerKey: PrivateKey, val request: OfferTypes.InvoiceRequest, val createAtSeconds: Long)
+
+class OfferManager(val nodeParams: NodeParams, val walletParams: WalletParams, val logger: MDCLogger) {
+    val remoteNodeId: PublicKey = walletParams.trampolineNode.id
+    private val pendingInvoiceRequests: HashMap<ByteVector32, PendingInvoiceRequest> = HashMap()
+    private val localOffers: HashMap<ByteVector32, OfferTypes.Offer> = HashMap()
+
+    fun registerOffer(offer: OfferTypes.Offer, pathId: ByteVector32) {
+        localOffers[pathId] = offer
+    }
+
+    fun requestInvoice(payOffer: PayOffer): List<OnionMessage> {
+        val payerKey = randomKey()
+        val request = OfferTypes.InvoiceRequest(payOffer.offer, payOffer.amount, payOffer.quantity, nodeParams.features.bolt12Features(), payerKey, nodeParams.chainHash)
+        val replyPathId = randomBytes32()
+        pendingInvoiceRequests[replyPathId] = PendingInvoiceRequest(payOffer, payerKey, request, currentTimestampSeconds())
+        val numHopsToAdd = max(0, payOffer.minReplyPathHops - 1)
+        val replyPathHops = (listOf(remoteNodeId) + List(numHopsToAdd) { nodeParams.nodeId }).map { IntermediateNode(it) }
+        val lastHop = Destination.Recipient(nodeParams.nodeId, replyPathId)
+        val replyPath = OnionMessages.buildRoute(randomKey(), replyPathHops, lastHop)
+        val messageContent = TlvStream(OnionMessagePayloadTlv.ReplyPath(replyPath), OnionMessagePayloadTlv.InvoiceRequest(request.records))
+        return payOffer.offer.contactInfos.mapNotNull { contactInfo -> buildMessage(randomKey(), randomKey(), listOf(IntermediateNode(remoteNodeId)), Destination(contactInfo), messageContent).right }
+    }
+
+    fun receiveMessage(msg: OnionMessage, channels: Map<ByteVector32, ChannelState>, currentBlockHeight: Int): OnionMessageAction? {
```

Comment:
It's overkill to pass the full channels map, the only thing you need is the list of remote `channel_update`s.